src/utils/keyAndAdresses.ts

```diff
@@ -0,0 +1,114 @@
+import {
+  base58Decode,
+  base58Encode,
+  varintEncode,
+  varintDecode,
+  hashBlake3,
+} from './Xbqcrypto';
+
+import { getBytesSecretKey, getBytesPublicKey } from './bytes';
+
+/**
+ * Prefixes for secret and public keys.
+ * Prefixes are used as a convention to differentiate one key from another.
+ */
+
+const PUBLIC_KEY_PREFIX = 'P';
+const ADDRESS_PREFIX = 'AU';
+const SECRET_KEY_PREFIX = 'S';
+
+/**
+ * A secret key.
+ *
+ * @remarks the secret key object is created from a base58 encoded string representing the secret key.
+ * The secret key bytes representation is `version + secretKey` where secretKey is the 32 bytes of the secret key.
+ */
+export class SecretKey {
+  base58Encoded: string;
+  base58Decoded: Uint8Array;
+  base58DecodedUnversioned: Uint8Array;
+  prefix: string;
+  version: number;
+
+  constructor(secretKeyBase58Encoded: string) {
+    this.prefix = SECRET_KEY_PREFIX;
+    this.base58Encoded = secretKeyBase58Encoded;
+    this.base58Decoded = getBytesSecretKey(secretKeyBase58Encoded);
+
+    // Slice off the version byte
+    this.base58DecodedUnversioned = this.base58Decoded.slice(1);
+
+    const secretKeyUnprefixed = secretKeyBase58Encoded.slice(
+      SECRET_KEY_PREFIX.length,
+    );
+    const decoded = base58Decode(secretKeyUnprefixed);
+    const decodedVersion = varintDecode(decoded);
+
+    // Extract version from the secret key
+    this.version = decodedVersion.value;
+  }
+}
+
+/**
+ * A public key.
+ *
+ * @remarks the public key object is created from a base58 encoded string representing the public key.
+ * The public key bytes representation is `version + publicKey` where publicKey is the 32 bytes of the public key.
+ *
+ * @remarks the public key is derived from the secret key and got the same version as the secret key.
+ */
+export class PublicKey {
+  version: number;
+  publicKey: Uint8Array;
+  base58Encoded: string;
+  base58Decoded: Uint8Array;
+  base58DecodedUnversioned: Uint8Array;
+
+  constructor(publicKeyArray: Uint8Array, secretKey: SecretKey) {
```

Comment:
If the public key is derived fro the privatekey, why do you set a publickeyarray argument ?

Reply:
Here is the process to create a publicKey object:
```
const publicKeyArray: Uint8Array = await ed.getPublicKey(base58DecodedUnversioned);
const publicKey: PublicKey = new PublicKey(publicKeyArray, secretKey);
```
We derive the publicKey array bytes with an async function, that's why I didn't want to put it in the constructor. The secretKey arg in the constructor is only here to give the right version.
Maybe we should make a custom async constructor and put everything in it ? Thus there would be only the secretKey argument.

Reply:
An async constructor feel weird. Can you put it in a FromPrivateKey function instead?
I still don't understand why the publicKeyArray is needed